[RC.x] Injecting missing import throws Cannot read property query of null

Hard to explain in the title

**Steps to reproduce and a minimal demo of the problem**

See http://plnkr.co/edit/BpjzGxQU5CEStzmWdWfh?p=preview , note that I correctly imported Renderer, but I forgot to add ElementRef. Renderer is undefined when it shouldn't.


**What steps should we try in your demo to see the problem?**

Run the plnkr and see the console

**Current behavior**

Now if I inject something I forgot to import no error is thrown and all of the classes injected in the constructor are undefined, even if they're correctly imported

**Expected/desired behavior**

It should work as <= b.15 throwing an error like

```
Cannot resolve all parameters for 'App'(Renderer, ?)
```
See plnkr http://plnkr.co/edit/aP6UWXsDOYaz3yV9U7Pp?p=preview with b.15

**Other information**

@ericmartinezr I'm trying to reproduce this under non-plunker conditions (ng2 test suite) and it is super-hard since if you compile things with TS you are going to get a very descriptive error: `error TS2304: Cannot find name 'ElementRef'`.

So I'm starting to suspect that this is more of a TS + SystemJS thing. I mean: it should have been reported by SystemJS. 

I will try some more to reproduce this under non-plunker conditions but it would be interesting to check with you / anyone else if you've ever seen this error outside plunks / SystemJS?

--- a/modules/@angular/compiler/src/metadata_resolver.ts
+++ b/modules/@angular/compiler/src/metadata_resolver.ts
@@ -260,11 +260,12 @@ export class CompileMetadataResolver {
 
   getDependenciesMetadata(typeOrFunc: Type | Function,
                           dependencies: any[]): cpl.CompileDiDependencyMetadata[] {
+    let hasUnknownDeps = false;
     let params = isPresent(dependencies) ? dependencies : this._reflector.parameters(typeOrFunc);
     if (isBlank(params)) {
       params = [];
     }
-    return params.map((param) => {
+    let dependenciesMetadata: cpl.CompileDiDependencyMetadata[] = params.map((param) => {
       if (isBlank(param)) {
         return null;
       }
@@ -306,6 +307,7 @@ export class CompileMetadataResolver {
         token = param;
       }
       if (isBlank(token)) {
+        hasUnknownDeps = true;
         return null;
       }
       return new cpl.CompileDiDependencyMetadata({
@@ -320,6 +322,15 @@ export class CompileMetadataResolver {
       });
 
     });
+
+    if (hasUnknownDeps) {
+      let depsTokens = dependenciesMetadata.map((dep) => {
+        return dep ? stringify(dep.token) : '?';
+      }).join(', ');
+      throw new BaseException(`Can't resolve all parameters for ${stringify(typeOrFunc)}: (${depsTokens}).`);
+    }
+
+    return dependenciesMetadata;
   }
 
   getTokenMetadata(token: any): cpl.CompileTokenMetadata {

--- a/modules/@angular/compiler/test/metadata_resolver_spec.ts
+++ b/modules/@angular/compiler/test/metadata_resolver_spec.ts
@@ -78,6 +78,14 @@ export function main() {
                  .toThrowError(`Expected 'styles' to be an array of strings.`);
            }
          }));
+
+      it('should throw with descriptive error message when provider token can not be resolved',
+         inject([CompileMetadataResolver], (resolver: CompileMetadataResolver) => {
+           if (!IS_DART) {
+             expect(() => resolver.getDirectiveMetadata(MyBrokenComp1))
+                 .toThrowError(`Can't resolve all parameters for MyBrokenComp1: (?).`);
+           }
+         }));
     });
 
     describe('getViewDirectivesMetadata', () => {
@@ -147,4 +155,9 @@ class ComponentWithEverything implements OnChanges,
   ngAfterContentChecked(): void {}
   ngAfterViewInit(): void {}
   ngAfterViewChecked(): void {}
+}
+
+@Component({selector: 'my-broken-comp', template: ''})
+class MyBrokenComp1 {
+  constructor(public dependency: any) {}
 }

